### PR TITLE
Enhance getInputProps to allow passing of non-overridden props

### LINKED ISSUE #34996: [Enhancement]: Form component input props should allow passing of custom callbacks
### Describe the solution you'd like

For the `Form` component `getInputProps`, we occasionally have to override an event like `onBlur`.

```
<TextControl
  { ...getInputProps( 'field_name' ) }
  onBlur={ myOnBlurHandler }
/>
```

The problem is that this overrides the default `getInputProps` `onBlur` method and now manually needs to be called:

```
<TextControl
  { ...getInputProps( 'field_name' ) }
  onBlur={ () => {
    myOnBlurHandler();
    getInputProps( 'field_name' ).onBlur();
  }
/>
```

To make things a bit easier on consumers, we could consider an optional second argument that adds props to the event without overriding the `Form`'s handlers.  E.g.,

```
<TextControl
  { ...getInputProps( 'field_name', { onBlur: myOnBlurHandler } ) }
/>
```

### Describe alternatives you've considered

Manually calling the input form's handler is the current way of handling this, but it is more prone to errors and more cumbersome to write.

`getInputProps( 'field_name' ).onBlur();`

### Additional context

The new product management experience has been creating utils like `getTextControlProps` and `getInputControlProps` that handle specific items like classes.  It would be good to move some of this logic back to the `Form` package.

### PR DESCRIPTION
**Pull request by joshuatf**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Previously, adding an `onChange` handler or anything else that collided with `getInputProps` would require recalling `getInputProps.someFunction` inside the new handler.  This PR introduces a second argument to `getInputProps` which passes props to the control without overriding the original handlers where possible.

Some props will not override the original handler (e.g., if you pass on `onChange` it will be combined with the `Form` component's `onChange`) and some will override (e.g., `value` can not be combined and the consumer prop will override the original value).

This PR also uncovered numerous TS issues, so a `getCheckboxProps` was introduced into the `Form` component to omit invalid props (e.g., `value` for checkboxes).  More specific prop getters are probably in order (`getSelectProps`), but I think we can iterate on these as we go.

Finally, a `sanitize` method was added as part of the optional consumer input props.  This is a shortcut to sanitizing the input value and handling changes on blur.

**TL;DR:**

* `getInputProps` now takes a second argument to combine/override props where necessary
* `sanitize` method can be added to an input to handle sanitization on blur
* `getCheckboxProps` was introduced for props specific to checkboxes

Closes #34996 .

### How to test the changes in this Pull Request:

1. Navigate to the new product page `Products -> Add new (MVP)`
2. Check that styling for checkboxes is still correct
3. Add values to the price and dimension fields, using thousand commas and decimal separators (depending your locale)
4. Make sure they are sanitized on blur and persist correctly on save

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce-admin/client/products/sections/pricing-section.tsx [hunk 1]
```diff
@@ -135,10 +132,8 @@ export const PricingSection: React.FC = () => {
 							{ ...regularPriceProps }
 							label={ __( 'List price', 'woocommerce' ) }
 							placeholder={ __( '10.59', 'woocommerce' ) }
-							onChange={ ( value: string ) => {
-								const sanitizedValue = sanitizePrice( value );
-								regularPriceProps?.onChange( sanitizedValue );
-							} }
+							value={ formatCurrencyDisplayValue( regularPriceProps?.value, currencyConfig, formatAmount ) }
+							onBlur={ () => setValue( 'regular_price', sanitizePrice( regularPriceProps.value ) ) }
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/sections/pricing-section.tsx]: @octaedro I moved the sanitization to `onBlur`.  I think this is a little bit more conventional as I was worried about input being sanitized as it was typed could lead to odd behavior.

But please let me know if I broke anything here or if this needs to be sanitized as we type.

File: plugins/woocommerce-admin/client/products/sections/utils.ts [hunk 2]
```diff
@@ -1,110 +1,79 @@
...
+ * @param {object}  context - Currency context
+ * @return {object} Props.
+ */
+export const getCurrencyInputProps = ( currencyConfig: CurrencyConfig ) => {
+	const { symbol, symbolPosition } = currencyConfig;
+	const currencyPosition = symbolPosition.includes( 'left' )
+		? 'prefix'
+		: 'suffix';
+
 	return {
```
Comment (joshuatf) [inline plugins/woocommerce-admin/client/products/sections/utils.ts:55]: My goal was to unwrap some of these so that we can:
1. More easily follow where props are being passed
2. Reduce the knowledge of form handlers in this function
3. Reduce overall logic in this function

**Review by joshuatf — COMMENTED** [review]
Just noting that we have more upcoming work (I also have an open PR) that will benefit from being able to pass in additional handlers without overriding the `Form`'s handlers.

File: plugins/woocommerce-admin/client/products/sections/pricing-section.tsx (hunk 1 above)
Comment (octaedro) [reply plugins/woocommerce-admin/client/products/sections/pricing-section.tsx]: Sounds good to me! Good idea!

**Comment by joshuatf** [comment]
@octaedro Thanks for the ping on rebasing this one in Slack!  After rebasing, a lot of TS issues cropped up that I think we were able to errantly bypass before.

I added a `getCheckboxProps` method to `Form` and also introduced a `sanitize` method as a shortcut for inputs.  I updated the description above to reflect all the changes.

@mdperez86 since this also affects a lot of your recent additions around dimensions, it would be great to get your 👀  on this as well.

File: packages/js/components/src/form/form.tsx [hunk 3]
```diff
@@ -268,35 +297,59 @@ function FormComponent< Values extends Record< string, any > >(
...
+			},
+			className: classnames( classNameProp, {
+				'has-error': isTouched && inputError,
+			} ),
 			help: isTouched ? ( inputError as string ) : null,
+			...additionalProps,
 		};
 	}
 
+	function getCheckboxProps< Value = Values[ keyof Values ] >(
```
Comment (joshuatf) [inline packages/js/components/src/form/form.tsx]: Should this be `getCheckboxInputProps` or `getCheckboxControlProps` instead?  The reason I ask is we'll probably want to introduce a `getSelectProps` which sounds odd and could be interpreted the wrong way.  Also okay leaving it as is, but wanted to get some feedback.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx [hunk 4]
```diff
@@ -78,13 +84,9 @@ export const ProductDetailsSection: React.FC = () => {
...
-							{ ...getTextControlProps(
-								getInputProps( 'name' )
-							) }
-							onBlur={ () => {
-								setSkuIfEmpty();
-								getInputProps( 'name' ).onBlur();
-							} }
+							{ ...getInputProps( 'name', {
+								onBlur: setSkuIfEmpty,
+							} ) }
```
Comment (octaedro) [inline plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:101]: ❤️

File: packages/js/components/src/form/form.tsx (hunk 3 above)
Comment (octaedro) [reply packages/js/components/src/form/form.tsx]: I would lean towards `getCheckboxControlProps`. But I'm also ok leaving it as it is now.

**Review by octaedro — COMMENTED** [review]
Nice job @joshuatf! This is testing well here and the code looks good.
There is a small visual error with the `Feature this product` checkbox, the rest looks great!

Screen Shot 2022-10-19 at 09 55 35

File: packages/js/components/src/form/form.tsx (hunk 3 above)
Comment (joshuatf) [reply packages/js/components/src/form/form.tsx]: I think that makes sense too since we're really creating props specifically for our controls 👍

**Comment by joshuatf** [comment]
>There is a small visual error with the Feature this product checkbox, the rest looks great!

Good catch @octaedro!  Updated in https://github.com/woocommerce/woocommerce/pull/35034/commits/38f80d8cca8ec53ef5b98000b6cb82fe8db64b29

Also gave this a rebase; could you take a look when you have time?

**Review by octaedro — DISMISSED** [review]
Thank you @joshuatf for addressing the changes! The `initialValues` issue was a pretty tricky thing to deal with.
Good job! :shipit:

**Review by octaedro — APPROVED** [review]
🚀
